odbc/src/api/odbc_uri.rs

```diff
@@ -0,0 +1,201 @@
+                }
+                // ODBC attribute keys are case insensitive, so we lowercase the keys
+                Ok((sp.remove(0).to_lowercase(), sp.remove(0)))
+            })
+            .collect::<Result<BTreeMap<_, _>>>()
+            .map(ODBCUri)
+    }
+
+    // remove will remove the first value with a given one of the names passed, assuming all names
+    // are synonyms.
+    pub fn remove(&mut self, names: &[&str]) -> Option<&'a str> {
+        for name in names.iter() {
+            let ret = self.0.remove(&name.to_string());
+            if ret.is_some() {
+                return ret;
+            }
+        }
+        None
+    }
+
+    // remove_or_else is the same as remove but with a default thunk.
+    pub fn remove_or_else(&mut self, f: impl Fn() -> &'a str, names: &[&str]) -> &'a str {
+        self.remove(names).unwrap_or_else(f)
+    }
+
+    // remove_mandatory_attribute will find an attribute that must exist and transfer ownership to
+    // the caller. It accepts a slice of names that will be checked in order for names that are
+    // synonyms (e.g., uid and user are both viable attribute names for a user). If both names
+    // exist, it will only find the first.
+    fn remove_mandatory_attribute(&mut self, names: &[&str]) -> Result<&'a str> {
+        self.remove(names).ok_or_else(|| {
+            if names.len() == 1 {
+                ODBCError::InvalidUriFormat(format!(
+                    "{} is required for a valid Mongo ODBC Uri",
+                    names[0]
+                ))
+            } else {
+                ODBCError::InvalidUriFormat(format!(
+                    "One of {:?} is required for a valid Mongo ODBC Uri",
+                    names
+                ))
+            }
+        })
+    }
+
+    // remove_to_mongo_uri converts this ODBCUri to a mongo_uri String. It will
+    // remove all the attributes necessary to make a mongo_uri. This is destructive!
+    pub fn remove_to_mongo_uri(&mut self) -> Result<String> {
+        let user = self.remove_mandatory_attribute(USER)?;
+        let pwd = self.remove_mandatory_attribute(PWD)?;
+        // TODO SQL-990: Support the PORT attribute, right now the only way to specify PORT is as
+        // part of SERVER. If ports are specified in both SERVER and PORT and they do not match it
+        // should be an error (I think, check the spec if it says...).
+        let server = self.remove_mandatory_attribute(SERVER)?;
+        let ssl = self.remove(SSL);
+        // TODO SQL-990: we may wish to support more attributes as options.
+        // If we do, add more tests to cover them.
+        let ssl_string =
+            if ssl.is_some() && ssl.unwrap() != "0" && ssl.unwrap().to_lowercase() != "false" {
+                "?ssl=true"
+            } else {
+                ""
+            };
+        Ok(format!(
+            "mongodb://{}:{}@{}{}",
+            user, pwd, server, ssl_string
+        ))
+    }
+}
+
+mod unit {
```

Comment:
I'm wondering if individually named tests would be better than two functions with many assertions each. I know line numbers will help us sort out what's failing, but doesn't this mean we'll need to run the tests sequentially to see all failures? By that, I mean if an assertion on line 112 fails, we won't see if the assertions after that pass or fail since the test will terminate. That might not lend itself to easiest maintenance in the future.

Reply:
Couldn't hurt.

Reply:
I also changed the name of `tests.rs` to `unit.rs` and removed the sub `unit` module. All the tests in there are unit tests, it was redundant. 
For clarity:
We need `integration` and `unit` in every test name because we now use `cargo test unit` and `cargo test integration` to run those separately in different evergreen tasks.

It's something we might want to consider for mongosql-rs at some point, too. But we already have the tests that run separately because they require go.